File: lib/battle/world/input_log.dart
import 'dart:convert';
import 'dart:typed_data';

import 'battle_input.dart';

/// 03_BATTLE_ENGINE.md §14. 전투 1회의 재생 가능한 전체 기록.
///
/// 리플레이 = `BattleWorld(config, seed)` 새로 만들고 [inputs]를 tick
/// 오름차순으로 다시 enqueue하는 것 — 이 클래스 자체는 순수 데이터라
/// 시뮬레이션 로직을 갖지 않는다.
class InputLog {
  const InputLog({
    required this.seed,
    required this.dataVersion,
    required this.stageId,
    required this.inputs, // tick 오름차순
    required this.formationHash,
  });

  final int seed;
  final String dataVersion;
  final String stageId;
  final List<BattleInput> inputs;
  final String formationHash;

  /// 서버 제출용 압축 인코딩 (varint delta). tick은 오름차순이 전제라
  /// 항상 음이 아닌 델타만 나온다 — seed만 zigzag(음수 가능)를 쓴다.
  Uint8List encode() {
    final b = BytesBuilder();
    _writeVarint(b, _zigzag(seed));
    _writeString(b, dataVersion);
    _writeString(b, stageId);
    _writeString(b, formationHash);
    _writeVarint(b, inputs.length);

    var prevTick = 0;
    for (final input in inputs) {
      _writeVarint(b, input.tick - prevTick);
      prevTick = input.tick;
      switch (input) {
        case SummonInput(:final slotIndex):
          b.addByte(0);
          _writeVarint(b, slotIndex);
        case UltimateInput():
          b.addByte(1);
        case FocusBoostInput(:final stage):
          b.addByte(2);
          _writeVarint(b, stage);
        case PageSwitchInput(:final page):
          b.addByte(3);
          _writeVarint(b, page);
      }
    }
    return b.toBytes();
  }

  /// [encode]의 역.
  static InputLog decode(Uint8List bytes) {
    final r = _ByteReader(bytes);
    final seed = _unzigzag(r.readVarint());
    final dataVersion = r.readString();
    final stageId = r.readString();
    final formationHash = r.readString();
    final count = r.readVarint();

    var tick = 0;
    final inputs = <BattleInput>[];
    for (var i = 0; i < count; i++) {
      tick += r.readVarint();
      final type = r.readByte();
      inputs.add(
        switch (type) {
          0 => SummonInput(tick, r.readVarint()),
          1 => UltimateInput(tick),
          2 => FocusBoostInput(tick, r.readVarint()),
          3 => PageSwitchInput(tick, r.readVarint()),
          _ => throw FormatException('알 수 없는 BattleInput 타입: $type'),
        },
      );
    }
    return InputLog(
      seed: seed,
      dataVersion: dataVersion,
      stageId: stageId,
      inputs: inputs,
      formationHash: formationHash,
    );
  }
}

int _zigzag(int v) => v >= 0 ? v * 2 : -v * 2 - 1;
int _unzigzag(int v) => v.isEven ? v ~/ 2 : -(v + 1) ~/ 2;

void _writeVarint(BytesBuilder b, int value) {
  var v = value;
  while (true) {
    final byte = v & 0x7f;
    v >>= 7;
    if (v == 0) {
      b.addByte(byte);
      return;
    }
    b.addByte(byte | 0x80);
  }
}

void _writeString(BytesBuilder b, String s) {
  final bytes = utf8.encode(s);
  _writeVarint(b, bytes.length);
  b.add(bytes);
}

class _ByteReader {
  _ByteReader(this._bytes);
  final Uint8List _bytes;
  int _pos = 0;

  int readByte() => _bytes[_pos++];

  int readVarint() {
    var result = 0;
    var shift = 0;
    while (true) {
      final byte = _bytes[_pos++];
      result |= (byte & 0x7f) << shift;
      if (byte & 0x80 == 0) return result;
      shift += 7;
    }
  }

  String readString() {
    final len = readVarint();
    final s = utf8.decode(_bytes.sublist(_pos, _pos + len));
    _pos += len;
    return s;
  }
}

File: functions/src/battle/inputLog.ts
/**
 * lib/battle/world/input_log.dart 의 varint-delta 인코딩을 그대로 포팅한 것.
 * V11(inputLog 디코딩 성공, tick 단조 증가) 검증과 V7/V9/V10이 로그 내용을
 * 들여다볼 수 있어야 해서 서버에도 같은 디코더가 필요하다.
 */
export type BattleInput =
  | { type: 'SUMMON'; tick: number; slotIndex: number }
  | { type: 'ULTIMATE'; tick: number }
  | { type: 'FOCUS_BOOST'; tick: number; stage: number }
  | { type: 'PAGE_SWITCH'; tick: number; page: number };

export interface DecodedInputLog {
  seed: number;
  dataVersion: string;
  stageId: string;
  formationHash: string;
  inputs: BattleInput[];
}

function zigzag(v: number): number {
  return v >= 0 ? v * 2 : -v * 2 - 1;
}

function unzigzag(v: number): number {
  return v % 2 === 0 ? v / 2 : -(v + 1) / 2;
}

function writeVarint(bytes: number[], value: number): void {
  let v = value;
  while (true) {
    const byte = v & 0x7f;
    v = Math.floor(v / 128);
    if (v === 0) {
      bytes.push(byte);
      return;
    }
    bytes.push(byte | 0x80);
  }
}

function writeString(bytes: number[], s: string): void {
  const buf = Buffer.from(s, 'utf8');
  writeVarint(bytes, buf.length);
  for (const b of buf) bytes.push(b);
}

export function encodeInputLog(log: DecodedInputLog): Buffer {
  const bytes: number[] = [];
  writeVarint(bytes, zigzag(log.seed));
  writeString(bytes, log.dataVersion);
  writeString(bytes, log.stageId);
  writeString(bytes, log.formationHash);
  writeVarint(bytes, log.inputs.length);

  let prevTick = 0;
  for (const input of log.inputs) {
    writeVarint(bytes, input.tick - prevTick);
    prevTick = input.tick;
    switch (input.type) {
      case 'SUMMON':
        bytes.push(0);
        writeVarint(bytes, input.slotIndex);
        break;
      case 'ULTIMATE':
        bytes.push(1);
        break;
      case 'FOCUS_BOOST':
        bytes.push(2);
        writeVarint(bytes, input.stage);
        break;
      case 'PAGE_SWITCH':
        bytes.push(3);
        writeVarint(bytes, input.page);
        break;
    }
  }
  return Buffer.from(bytes);
}

class ByteReader {
  private pos = 0;
  constructor(private readonly bytes: Buffer) {}

  readByte(): number {
    if (this.pos >= this.bytes.length) throw new Error('inputLog: unexpected end of buffer');
    return this.bytes[this.pos++];
  }

  readVarint(): number {
    // 32비트 시프트(<<)는 2^31을 넘는 값(예: zigzag된 32bit seed)에서 부호
    // 비트로 깨지므로, 시프트 대신 배수 누적으로 안전 정수 범위까지 다룬다.
    let result = 0;
    let shift = 0;
    while (true) {
      const byte = this.readByte();
      result += (byte & 0x7f) * 2 ** shift;
      if ((byte & 0x80) === 0) return result;
      shift += 7;
    }
  }

  readString(): string {
    const len = this.readVarint();
    if (this.pos + len > this.bytes.length) throw new Error('inputLog: string length out of range');
    const s = this.bytes.subarray(this.pos, this.pos + len).toString('utf8');
    this.pos += len;
    return s;
  }

  get remaining(): number {
    return this.bytes.length - this.pos;
  }
}

/** V11: 디코딩 실패, tick 역행, endTick 초과는 전부 예외로 던진다. */
export function decodeInputLog(base64: string, endTick: number): DecodedInputLog {
  const buf = Buffer.from(base64, 'base64');
  const r = new ByteReader(buf);
  const seed = unzigzag(r.readVarint());
  const dataVersion = r.readString();
  const stageId = r.readString();
  const formationHash = r.readString();
  const count = r.readVarint();

  let tick = 0;
  const inputs: BattleInput[] = [];
  for (let i = 0; i < count; i++) {
    const delta = r.readVarint();
    if (delta < 0) throw new Error('inputLog: negative tick delta');
    tick += delta;
    if (tick > endTick) throw new Error(`inputLog: tick ${tick} exceeds endTick ${endTick}`);
    const type = r.readByte();
    switch (type) {
      case 0:
        inputs.push({ type: 'SUMMON', tick, slotIndex: r.readVarint() });
        break;
      case 1:
        inputs.push({ type: 'ULTIMATE', tick });
        break;
      case 2:
        inputs.push({ type: 'FOCUS_BOOST', tick, stage: r.readVarint() });
        break;
      case 3:
        inputs.push({ type: 'PAGE_SWITCH', tick, page: r.readVarint() });
        break;
      default:
        throw new Error(`inputLog: unknown BattleInput type ${type}`);
    }
  }
  if (r.remaining !== 0) throw new Error('inputLog: trailing bytes after decoding');

  return { seed, dataVersion, stageId, formationHash, inputs };
}
